File: skills/ultracode/scripts/new-ultracode-run.py
#!/usr/bin/env python3
"""Create an Ultracode workflow run directory."""

from __future__ import annotations

import argparse
import json
import re
import sys
from datetime import datetime, timedelta, timezone
from pathlib import Path


GitContext = tuple[Path, Path]


def slugify(value: str) -> str:
    slug = re.sub(r"[^a-zA-Z0-9]+", "-", value.lower()).strip("-")
    return slug[:48].strip("-") or "workflow"


def write_if_missing(path: Path, content: str) -> None:
    if path.exists():
        raise FileExistsError(f"{path} already exists")
    path.write_text(content, encoding="utf-8")


def git_context(root: Path) -> GitContext | None:
    for candidate in (root, *root.parents):
        dot_git = candidate / ".git"
        if dot_git.is_dir():
            return candidate, dot_git

        if dot_git.is_file():
            content = dot_git.read_text(encoding="utf-8", errors="replace").strip()
            prefix = "gitdir:"
            if content.lower().startswith(prefix):
                git_dir = Path(content[len(prefix) :].strip())
                if not git_dir.is_absolute():
                    git_dir = (candidate / git_dir).resolve()
                return candidate, git_dir

    return None


def workflow_exclude_pattern(root: Path, worktree_root: Path) -> str:
    workflow_dir = (root / ".workflow").resolve()
    try:
        relative = workflow_dir.relative_to(worktree_root.resolve())
    except ValueError:
        return ".workflow/"
    return f"{relative.as_posix().rstrip('/')}/"


def ensure_workflow_git_excluded(root: Path) -> str:
    context = git_context(root)
    if context is None:
        return "not-git"

    worktree_root, git_dir = context
    exclude_path = git_dir / "info" / "exclude"
    pattern = workflow_exclude_pattern(root, worktree_root)
    exclude_path.parent.mkdir(parents=True, exist_ok=True)
    existing = exclude_path.read_text(encoding="utf-8") if exclude_path.exists() else ""
    patterns = {line.strip() for line in existing.splitlines() if line.strip() and not line.lstrip().startswith("#")}
    if pattern in patterns or pattern.rstrip("/") in patterns:
        return "already-excluded"

    prefix = "" if not existing or existing.endswith("\n") else "\n"
    with exclude_path.open("a", encoding="utf-8") as handle:
        handle.write(f"{prefix}{pattern}\n")
    return "added-local-exclude"


def main() -> int:
    parser = argparse.ArgumentParser(description="Create an Ultracode workflow run directory.")
    parser.add_argument("title", help="Short task title")
    parser.add_argument("--root", default=".", help="Project root where .workflow should be created")
    parser.add_argument("--slug", help="Optional slug override")
    parser.add_argument("--mode", choices=["workflow", "delegated", "runner"], default="delegated")
    parser.add_argument("--no-git-exclude", action="store_true", help="Do not add .workflow/ to local Git exclude")
    args = parser.parse_args()

    root = Path(args.root).expanduser().resolve()
    if not args.no_git_exclude:
        exclude_status = ensure_workflow_git_excluded(root)
        if exclude_status == "added-local-exclude":
            print("Added workflow artifact path to local Git exclude", file=sys.stderr)
        elif exclude_status == "not-git":
            print("Warning: root is not a Git checkout; keep .workflow/ out of version control", file=sys.stderr)

    now = datetime.now(timezone.utc)
    next_check_in = now + timedelta(minutes=10)
    run_slug = slugify(args.slug or args.title)
    run_id = f"{now.strftime('%Y%m%d-%H%M%S')}-{run_slug}"
    run_dir = root / ".workflow" / "ultracode" / run_id

    if run_dir.exists():
        raise SystemExit(f"Run directory already exists: {run_dir}")

    (run_dir / "packets").mkdir(parents=True)
    (run_dir / "results").mkdir()

    state = {
        "title": args.title,
        "runId": run_id,
        "createdAt": now.isoformat().replace("+00:00", "Z"),
        "status": "planning",
        "mode": args.mode,
        "root": str(root),
        "checkInIntervalMinutes": 10,
        "lastCheckInAt": None,
        "nextCheckInDueAt": next_check_in.isoformat().replace("+00:00", "Z"),
        "checkIns": [],
        "resources": [],
        "packets": [],
        "approvals": [],
        "verification": [],
    }

    write_if_missing(
        run_dir / "state.json",
        json.dumps(state, indent=2) + "\n",
    )
    write_if_missing(
        run_dir / "plan.md",
        f"""# Plan

## Goal

{args.title}

## Success Criteria

- [ ] 

## Non-Goals

- 

## Constraints

- 

## Risks

- 

## Verification Gates

- [ ] 
""",
    )
    write_if_missing(
        run_dir / "orchestration.md",
        f"""# Orchestration

## Mode

{args.mode}

## Host Capabilities

- Native subagents:
- Runner:
- Network:
- Write access:

## Work Packets

| Packet | Owner | Scope | Dependencies | Expected result |
| --- | --- | --- | --- | --- |

## Check-In Cadence

- Interval: 10 minutes
- Last check-in:
- Next check-in: {next_check_in.isoformat().replace("+00:00", "Z")}
- Phase-change updates:

## Resource Plan

| Resource | Owner | Purpose | Cleanup |
| --- | --- | --- | --- |

## Stop Conditions

- 

## Budget

- Workers:
- Time:
- Token/cost limit:
""",
    )
    write_if_missing(
        run_dir / "integration.md",
        """# Integration

## Results Reviewed

- 

## Accepted

- 

## Rejected

- 

## Conflicts Resolved

- 

## Integrated Changes

- 
""",
    )
    write_if_missing(
        run_dir / "final-report.md",
        """# Final Report

## Outcome

## Files Changed

- 

## Verification Evidence

- 

## Resource Cleanup

- 

## Remaining Risks

- 

## Follow-Ups

- 
""",
    )

    print(run_dir)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: extras/skills/ultracode/scripts/new-ultracode-run.mjs
#!/usr/bin/env node
import { appendFile, mkdir, readFile, stat, writeFile } from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';
import process from 'node:process';

const VALID_MODES = new Set(['durable', 'runner', 'workflow', 'delegated']);

function usage() {
  return `Usage: new-ultracode-run.mjs [options] "short task title"

Create an Ultracode workflow run directory.

Options:
  --root <path>          Project root where .workflow should be created (default: .)
  --slug <slug>          Optional slug override
  --mode <mode>          durable or runner (default: durable). workflow/delegated are accepted as durable aliases
  --no-git-exclude       Do not add .workflow/ to local Git exclude
  -h, --help             Show this help
`;
}

function parseArgs(argv) {
  const args = {
    root: '.',
    slug: null,
    mode: 'durable',
    noGitExclude: false,
    title: null,
  };

  const positionals = [];
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];

    if (arg === '-h' || arg === '--help') {
      args.help = true;
      continue;
    }

    if (arg === '--no-git-exclude') {
      args.noGitExclude = true;
      continue;
    }

    if (arg.startsWith('--root=')) {
      args.root = arg.slice('--root='.length);
      continue;
    }

    if (arg === '--root') {
      args.root = requireValue(argv, index, arg);
      index += 1;
      continue;
    }

    if (arg.startsWith('--slug=')) {
      args.slug = arg.slice('--slug='.length);
      continue;
    }

    if (arg === '--slug') {
      args.slug = requireValue(argv, index, arg);
      index += 1;
      continue;
    }

    if (arg.startsWith('--mode=')) {
      args.mode = arg.slice('--mode='.length);
      continue;
    }

    if (arg === '--mode') {
      args.mode = requireValue(argv, index, arg);
      index += 1;
      continue;
    }

    if (arg.startsWith('-')) {
      throw new Error(`Unknown option: ${arg}`);
    }

    positionals.push(arg);
  }

  if (positionals.length > 1) {
    throw new Error(`Expected one title, received ${positionals.length}`);
  }

  args.title = positionals[0] ?? null;

  if (!args.help && !args.title) {
    throw new Error('Missing required title');
  }

  if (!VALID_MODES.has(args.mode)) {
    throw new Error(`--mode must be one of: ${Array.from(VALID_MODES).join(', ')}`);
  }

  if (args.mode === 'workflow' || args.mode === 'delegated') {
    args.mode = 'durable';
  }

  return args;
}

function requireValue(argv, index, option) {
  const value = argv[index + 1];
  if (!value || value.startsWith('-')) {
    throw new Error(`${option} requires a value`);
  }
  return value;
}

function slugify(value) {
  const slug = value.toLowerCase().replace(/[^a-z0-9]+/gi, '-').replace(/^-+|-+$/g, '');
  return slug.slice(0, 48).replace(/-+$/g, '') || 'workflow';
}

function expandUser(value) {
  if (value === '~') {
    return os.homedir();
  }
  if (value.startsWith('~/')) {
    return path.join(os.homedir(), value.slice(2));
  }
  return value;
}

async function pathType(candidate) {
  try {
    const details = await stat(candidate);
    if (details.isDirectory()) return 'directory';
    if (details.isFile()) return 'file';
    return 'other';
  } catch (error) {
    if (error.code === 'ENOENT') return null;
    throw error;
  }
}

async function writeIfMissing(filePath, content) {
  if (await pathType(filePath)) {
    throw new Error(`${filePath} already exists`);
  }
  await writeFile(filePath, content, 'utf8');
}

async function gitContext(root) {
  let candidate = root;

  while (true) {
    const dotGit = path.join(candidate, '.git');
    const type = await pathType(dotGit);

    if (type === 'directory') {
      return { worktreeRoot: candidate, gitDir: dotGit };
    }

    if (type === 'file') {
      const content = (await readFile(dotGit, 'utf8')).trim();
      const prefix = 'gitdir:';
      if (content.toLowerCase().startsWith(prefix)) {
        let gitDir = content.slice(prefix.length).trim();
        if (!path.isAbsolute(gitDir)) {
          gitDir = path.resolve(candidate, gitDir);
        }
        return { worktreeRoot: candidate, gitDir };
      }
    }

    const parent = path.dirname(candidate);
    if (parent === candidate) {
      return null;
    }
    candidate = parent;
  }
}

function workflowExcludePattern(root, worktreeRoot) {
  const workflowDir = path.resolve(root, '.workflow');
  const relative = path.relative(path.resolve(worktreeRoot), workflowDir);

  if (!relative || relative.startsWith('..') || path.isAbsolute(relative)) {
    return '.workflow/';
  }

  return `${relative.split(path.sep).join('/').replace(/\/+$/g, '')}/`;
}

async function ensureWorkflowGitExcluded(root) {
  const context = await gitContext(root);
  if (!context) {
    return 'not-git';
  }

  const excludePath = path.join(context.gitDir, 'info', 'exclude');
  const pattern = workflowExcludePattern(root, context.worktreeRoot);
  await mkdir(path.dirname(excludePath), { recursive: true });

  let existing = '';
  if (await pathType(excludePath)) {
    existing = await readFile(excludePath, 'utf8');
  }

  const patterns = new Set(
    existing
      .split(/\r?\n/)
      .map((line) => line.trim())
      .filter((line) => line && !line.startsWith('#')),
  );

  if (patterns.has(pattern) || patterns.has(pattern.replace(/\/$/g, ''))) {
    return 'already-excluded';
  }

  const prefix = !existing || existing.endsWith('\n') ? '' : '\n';
  await appendFile(excludePath, `${prefix}${pattern}\n`, 'utf8');
  return 'added-local-exclude';
}

function pad(value) {
  return String(value).padStart(2, '0');
}

function formatRunTimestamp(date) {
  return [
    date.getUTCFullYear(),
    pad(date.getUTCMonth() + 1),
    pad(date.getUTCDate()),
    '-',
    pad(date.getUTCHours()),
    pad(date.getUTCMinutes()),
    pad(date.getUTCSeconds()),
  ].join('');
}

function isoZ(date) {
  return date.toISOString();
}

async function main() {
  let args;
  try {
    args = parseArgs(process.argv.slice(2));
  } catch (error) {
    console.error(`ERROR: ${error.message}`);
    console.error(usage().trimEnd());
    return 2;
  }

  if (args.help) {
    console.log(usage().trimEnd());
    return 0;
  }

  const root = path.resolve(expandUser(args.root));
  if (!args.noGitExclude) {
    const excludeStatus = await ensureWorkflowGitExcluded(root);
    if (excludeStatus === 'added-local-exclude') {
      console.error('Added workflow artifact path to local Git exclude');
    } else if (excludeStatus === 'not-git') {
      console.error('Warning: root is not a Git checkout; keep .workflow/ out of version control');
    }
  }

  const now = new Date();
  const runSlug = slugify(args.slug || args.title);
  const runId = `${formatRunTimestamp(now)}-${runSlug}`;
  const runDir = path.join(root, '.workflow', 'ultracode', runId);

  if (await pathType(runDir)) {
    throw new Error(`Run directory already exists: ${runDir}`);
  }

  await mkdir(path.join(runDir, 'packets'), { recursive: true });
  await mkdir(path.join(runDir, 'results'));

  const state = {
    title: args.title,
    runId,
    createdAt: isoZ(now),
    updatedAt: isoZ(now),
    status: 'planning',
    mode: args.mode,
    root,
    budget: {
      maxConcurrentAgents: 4,
      maxTotalAgents: 16,
      timeLimitMinutes: 45,
      network: false,
      credentialedTools: false,
    },
    approval: {
      status: 'pending',
      approvedAt: null,
      writes: [],
      verification: [],
    },
    progress: {
      percentComplete: 5,
      eta: 'unknown',
      etaConfidence: 'low',
      status: 'green',
      statusReason: 'Workflow initialized',
      done: [],
      remaining: ['Approve phase graph', 'Run phases', 'Reduce results', 'Verify'],
    },
    phases: [],
    agents: [],
    resources: [],
    verification: [],
    nativeWorkflow: null,
  };

  await writeIfMissing(path.join(runDir, 'state.json'), `${JSON.stringify(state, null, 2)}\n`);
  await writeIfMissing(
    path.join(runDir, 'workflow.md'),
    `# Workflow

## Goal

${args.title}

## Args

## Approval Envelope

- Mode: ${args.mode}
- Max concurrent agents: 4
- Max total agents: 16
- Time limit: 45 minutes
- Write scopes:
- Network: false
- Credentialed tools: false
- Verification gates:

## Phase Graph

| Phase | Depends On | Max Agents | Read Scope | Write Scope | Output |
| --- | --- | ---: | --- | --- | --- |

## Execution Rules

- Parent critical path:
- Worker packet rule:
- Reducer rule:
- Reviewer rule:
- Stop conditions:
`,
  );
  await writeIfMissing(
    path.join(runDir, 'journal.md'),
    `# Journal

## ${isoZ(now)}

- Workflow initialized.

Run directory: ${runDir}
`,
  );
  await writeIfMissing(
    path.join(runDir, 'integration.md'),
    `# Integration

## Results Reviewed

-

## Accepted

-

## Rejected

-

## Conflicts Resolved

-

## Integrated Changes

-

## Deferred

-
`,
  );
  await writeIfMissing(
    path.join(runDir, 'final-report.md'),
    `# Final Report

## Outcome

## Files Changed

-

## Verification Evidence

-

## Resource Cleanup

-

## Remaining Risks

-

## Follow-Ups

-
`,
  );

  console.log(runDir);
  return 0;
}

main()
  .then((code) => {
    process.exitCode = code;
  })
  .catch((error) => {
    console.error(`ERROR: ${error.message}`);
    process.exitCode = 1;
  });
